### PR TITLE
Fix C API to be compatible with latest 3.11 beta

### PR DESCRIPTION
**Pull request by albanD**
Based off https://github.com/pytorch/pytorch/pull/80511 with extra changes:
- Update pybind to the latest release as it contains some needed fixes
- Extend the compat header to do reduce changes in code

### CODE REVIEW

File: torch/csrc/autograd/functions/init.cpp [hunk 1]
```diff
@@ -161,8 +161,9 @@ void THPAutograd_initFunctions() {
   if (!c_module)
     throw python_error();
 
-  Py_INCREF(module);
+  Py_INCREF(module.get());
   if (PyModule_AddObject(c_module, "_functions", module) < 0) {
+    Py_DECREF(module.get());
```
Comment (albanD) [inline torch/csrc/autograd/functions/init.cpp:166]: This used to leak on error (very rare) and is now fixed.

**Review by malfet — APPROVED** [review]
LGTM, though this will likely increase the frequency of crash on shutdown issues

File: torch/csrc/Stream.cpp [hunk 2]
```diff
@@ -107,7 +107,7 @@ PyTypeObject THPStreamType = {
 
 void THPStream_init(PyObject* module) {
   THPStreamClass = &THPStreamType;
-  Py_TYPE(&THPStreamType) = &PyType_Type;
+  Py_SET_TYPE(&THPStreamType, &PyType_Type);
```
Comment (mattip) [inline torch/csrc/Stream.cpp:110]: [Py_SET_TYPE](https://docs.python.org/3/c-api/structures.html#c.Py_SET_TYPE) is documented for python 3.9 and up. 
```suggestion
#if PY_VERSION_HEX < 03090000
  Py_TYPE(&THPStreamType) = &PyType_Type;
#else
  Py_SET_TYPE(&THPStreamType, &PyType_Type);
#endif
```

Comment (mattip) [reply torch/csrc/Stream.cpp:110]: or is `torch/csrc/utils/python_compat.h` already included here somehow?

Comment (albanD) [reply torch/csrc/Stream.cpp:110]: `python_compat.h` is already included through `THP.h` -> `csrc/utils.h` indeed

**Review by mattip — APPROVED** [review]
LGTM. I like the use of the compat header to make the code cleaner. There are additional copies of `pybind11` in thirdparty submodules, at  different versions. Some are behind the 2.6.2 used here, will that cause problems for 3.11 support? Is there a reason this PR chose 2.6.2 (released Jan 2021) and not a later version?
```
$ git submodule status --recursive |grep pybind11
  third_party/onnx/third_party/pybind11 (v2.9.1)
  third_party/onnx-tensorrt/third_party/onnx/third_party/pybind11 (v2.0.0-353-ga104119)
  third_party/onnx-tensorrt/third_party/onnx/third_party/pybind11/tools/clang (6a00cbc)
  third_party/pybind11 (v2.6.2)
  third_party/tensorpipe/third_party/pybind11 (v2.2.4-1-ga23996f)
  third_party/tensorpipe/third_party/pybind11/tools/clang (6a00cbc)

```

**Comment by albanD** [comment]
Thanks a lot for the review @mattip 

> Some are behind the 2.6.2 used here, will that cause problems for 3.11 support?

That is a good question. Not that I know of since things do build properly locally and the tests run.
I will do a follow up PR to add 3.11 to the CI matrix and do testing, there might be stuff not caught locally that this will reveal.

> Is there a reason this PR chose 2.6.2 (released Jan 2021) and not a later version?

Not really beyond not moving too far if possible.
But I see that 2.10.0 was released 10 days ago and it adds official support for 3.11. So let me take a stab to move forward all the way to this one.

**Comment by mattip** [comment]
> it gets used preferably

Cool, I did not know that.

**Comment by robieta** [comment]
@albanD Unfortunately this more than doubles the overhead of the python profiler. (From ~100ns of overhead per fn call to ~225ns. For reference, a no-op function is ~75ns) The issue is the cache in `profiler_python.py` works really, really hard to avoid materializing data or making API calls unless absolutely necessary. And `CodeLocation` (which is on the hot path) has gone from a quite trivial struct to one that is doing non-trivial work.

That said, a lot of the changes are clearly quite good cleanups. Do you want to work together to recover the perf?

**Comment by albanD** [comment]
@robieta yes happy to look into that!
Could you share what you use to benchmark this?

Note that I am really not sure that the old version of CodeLocation was sound. In particular, it was storing a lot of PyObject* without ever doing refcounts bump on these object (or their parent).

In any case:
- We could relatively easily recover perf for the old python versions by keeping the same code as before.
- But for 3.11+, these field just don't exist anymore. So we will have to do this extra work. Or maybe just forget about CodeLocation altogether and just save the PyFrameObject* directly?

**Comment by robieta** [comment]
> Could you share what you use to benchmark this?

My benchmark was very simple:
```
import torch

def f():
  pass
  
%timeit f()
with torch.profiler.profile(with_stack=True):
  %timeit f()
```

> Note that I am really not sure that the old version of CodeLocation was sound. In particular, it was storing a lot of PyObject* without ever doing refcounts bump on these object (or their parent).

Yeah, so the deal there is that you're never allowed to actually use the stored object; it's purely for pointer identity. I agree that it's not 100% sound due to potential address reuse. (Though it could be made so with a gc callback)

> We could relatively easily recover perf for the old python versions by keeping the same code as before.

I think we should figure out a correct, reasonably future proof approach even if it means python tracing is slower in the mean time. But thanks for the offer.

> But for 3.11+, these field just don't exist anymore. So we will have to do this extra work.

So what changed exactly? Was it just some fields being shifted around, or was there a more fundamental change to the interpreter? 

> Or maybe just forget about CodeLocation altogether and just save the PyFrameObject* directly?

That won't work since I'm keying not on the particular call, but on the code that that frame is calling. (Also it's not ok to extend the lifetime of the frame.)

**Comment by albanD** [comment]
> So what changed exactly? Was it just some fields being shifted around, or was there a more fundamental change to the interpreter?

The whole PyFrameObject was moved and refactored. That means that you cannot access any field directly anymore, you have to use these accessor functions. And these functions, always return new references.
That is why I'm saving the strings/etc now because, to keep CodeLocation a trivial class I had to do that. Otherwise I would need a destructor to properly decreft all these objects it stores.

**Comment by robieta** [comment]
> PyFrameObject

Oh, interesting. Let me look over the release notes and come up with a tentative plan, and then I'll ping you to see if it seems reasonable.

**Pull request closed without merging**